lib/Conversion/LLHDToLLVM/LLHDToLLVM.cpp

```diff
@@ -0,0 +1,1464 @@
+// Unit conversions
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// Convert an `llhd.entity` unit to LLVM. The result is an `llvm.func` which
+/// takes a pointer to the state as arguments.
+struct EntityOpConversion : public ConvertToLLVMPattern {
+  explicit EntityOpConversion(MLIRContext *ctx,
+                              LLVMTypeConverter &typeConverter)
+      : ConvertToLLVMPattern(llhd::EntityOp::getOperationName(), ctx,
+                             typeConverter) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    // reset signal counter
+    signalCounter = 0;
+    // get adapted operands
+    EntityOpAdaptor transformed(operands);
+    // get entity operation
+    auto entityOp = cast<EntityOp>(op);
+
+    // collect llvm types
+    auto voidTy = getVoidType();
+    auto i8PtrTy = getVoidPtrType();
+    auto i32Ty = LLVM::LLVMType::getInt32Ty(&getDialect());
+
+    // have an intermediate signature conversion to add the arguments for the
+    // state, signal table and argument table poitner arguments
+    LLVMTypeConverter::SignatureConversion intermediate(
+        entityOp.getNumArguments());
+    // add state, signal table and arguments table arguments
+    intermediate.addInputs(
+        ArrayRef<Type>({i8PtrTy, i32Ty.getPointerTo(), i32Ty.getPointerTo()}));
+    for (size_t i = 0, e = entityOp.getNumArguments(); i < e; ++i)
+      intermediate.addInputs(i, voidTy);
+    rewriter.applySignatureConversion(&entityOp.getBody(), intermediate);
+
+    OpBuilder bodyBuilder =
+        OpBuilder::atBlockBegin(&entityOp.getBlocks().front());
+    LLVMTypeConverter::SignatureConversion final(
+        intermediate.getConvertedTypes().size());
+    final.addInputs(0, i8PtrTy);
+    final.addInputs(1, i32Ty.getPointerTo());
+    final.addInputs(2, i32Ty.getPointerTo());
+
+    for (size_t i = 0, e = entityOp.getNumArguments(); i < e; ++i) {
+      // create gep and load operations from arguments table for each original
+      // argument
+      auto index = bodyBuilder.create<LLVM::ConstantOp>(
+          rewriter.getUnknownLoc(), i32Ty, rewriter.getI32IntegerAttr(i));
+      auto bitcast = bodyBuilder.create<LLVM::GEPOp>(
+          rewriter.getUnknownLoc(), i32Ty.getPointerTo(),
+          entityOp.getArgument(2), ArrayRef<Value>(index));
+      auto load =
+          bodyBuilder.create<LLVM::LoadOp>(rewriter.getUnknownLoc(), bitcast);
+      // remap i-th original argument to the loaded value
+      final.remapInput(i + 3, load.getResult());
+    }
+
+    rewriter.applySignatureConversion(&entityOp.getBody(), final);
+
+    // converted entity signature
+    auto funcTy = LLVM::LLVMType::getFunctionTy(
+        voidTy, {i8PtrTy, i32Ty.getPointerTo(), i32Ty.getPointerTo()}, false);
+    // // create the llvm function
```

Comment:
nit: Can you fix this comment? Also, please use complete sentences and punctuation in comments.